crates/rpcsrv/src/coordinator/scheduler.rs

```diff
@@ -0,0 +1,209 @@
+            .await
+    }
+
+    /// Put a channel for receiving work.
+    pub async fn put_work_channel(&self, tx: oneshot::Sender<Option<StagedPlan>>) -> Result<()> {
+        self.send_msg(SchedulerMessage::PutWorkChannel { tx }).await
+    }
+
+    pub async fn shutdown(&self) -> Result<()> {
+        self.send_msg(SchedulerMessage::Shutdown).await
+    }
+
+    async fn send_msg(&self, msg: SchedulerMessage) -> Result<()> {
+        if self.send.send(msg).await.is_err() {
+            return Err(CoordinatorError::String(
+                "failed to send scheduled message".to_string(),
+            ));
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+enum PendingStream {
+    StreamArrivedFirst(ExecutionBatchStream),
+    WaitingForStream(oneshot::Sender<ExecutionBatchStream>),
+}
+
+#[derive(Debug)]
+pub struct Scheduler {
+    conf: SchedulerConfig,
+
+    send: mpsc::Sender<SchedulerMessage>,
+    recv: mpsc::Receiver<SchedulerMessage>,
+
+    /// Buffer of work yet to do be done.
+    ///
+    /// Storing these in a separate vec rather than in the channel lets us
+    /// implement prioritization and affinity tracking logic.
+    // TODO: Work item instead of just staged plan.
+    work_buf: VecDeque<StagedPlan>,
+
+    /// Track streams that have been sent back by workers, _or_ streams that the
+    /// client are waiting on.
+    pending_streams: HashMap<Uuid, PendingStream>,
+}
+
+/// Messages that can be sent to the scheduler.
+#[derive(Debug)]
+enum SchedulerMessage {
+    RegisterWorker {
+        node_id: Uuid,
+    },
+    DeregisterWorker {
+        node_id: Uuid,
+    },
+    SchedulePlan {
+        plan: StagedPlan,
+    },
+    PutBatchStreamChannel {
+        work_id: Uuid,
+        tx: oneshot::Sender<ExecutionBatchStream>,
+    },
+    PutWorkChannel {
+        tx: oneshot::Sender<Option<StagedPlan>>,
+    },
+    Shutdown,
+}
+
+impl Scheduler {
+    pub fn new() -> Self {
+        let conf = SchedulerConfig::default();
+        let (send, recv) = mpsc::channel(conf.chan_buffer);
+        Scheduler {
+            conf,
+            send,
+            recv,
+            work_buf: VecDeque::with_capacity(conf.work_buffer),
+            pending_streams: HashMap::new(),
+        }
+    }
+
+    pub fn client(&self) -> SchedulerClient {
+        SchedulerClient {
+            send: self.send.clone(),
+        }
+    }
+
+    pub async fn run(mut self) -> Result<()> {
+        debug!("starting scheduler");
+
+        let mut interval = tokio::time::interval(self.conf.debug_tick_interval);
+        loop {
+            tokio::select! {
+                _ = interval.tick() => {
+                    debug!("scheduler debug tick");
+                }
+
+                Some(msg) = self.recv.recv() => {
+                    if let SchedulerMessage::Shutdown = msg {
+                        // TODO: We could probably try draining pending work
+                        // before shutting down.
```

Comment:
yeah, it's hard to say, we don't have a lot of orchestration or management about clean/gentle shutdown at the moment. (e.g. we'd want to stop accepting new work but wait (a long time?) for pending work to return? seems like there's a fair piece of of machinery to orchestrate that and we can/should track this work somewhere.

Reply:
Added to future work here: https://github.com/GlareDB/glaredb/issues/1867